include/net/download_client.h

```diff
@@ -98,9 +100,41 @@ struct download_client_evt {
 };
 
 /**
- * @brief Download client configuration options.
+ * @brief Download client asynchronous event handler.
+ *
+ * Through this callback, the application receives events, such as
+ * download of a fragment, download completion, or errors.
+ *
+ * If the callback returns a non-zero value, the download stops.
+ * To resume the download, use @ref download_client_start().
+ *
+ * @param[in] event	The event.
+ *
+ * @return Zero to continue the download, non-zero otherwise.
+ */
+typedef int (*download_client_callback_t)(
+	const struct download_client_evt *event);
+
+/**
+ * @brief
  */
 struct download_client_cfg {
+	/** Event handler. */
+	download_client_callback_t callback;
+	/** Client buffer. */
+	char *buf;
+	/** Client buffer size. */
+	size_t buf_size;
+};
+
+/**
+ * @brief Download client configuration options.
+ */
+struct download_client_host_cfg {
+	/** Server hosting the file, null-terminated.
+	 *  The host name must be kept in scope while download is going on.
+	 */
+	const char *hostname;
```

Comment:
Since you are so heavily refactoring, I would propose that we get rid of this separation of host and path.
Just use URI on the API.
It is actually very rare that any interfaces would give us host and path separately. I only know nRF Cloud to do so on some REST APIs. It is more common that we receive full URI where to fetch the data.

Reply:
Oh.. and also now host is provided inside a struct and path is provided as a function parameter. This emphasizes the separation even further.

It feels like cleanest API would just take URI as one parameter, or field inside this struct. Let the HTTP parser to split protocol, host, port and path.